File: src/lib/aptams/tables.ts
// Loader for the national-standard scoring tables extracted from the official PDF.
// Port of aptams/rule_engine/tables.py + analysis/extract_scoring_tables.py.
//
// The single auditable origin of every scored number is:
//   data/ML/...大学生国家体质健康测试评分标准.pdf  (parsed by the extractor)
//     -> src/lib/aptams/data/university_2014.json   (loaded here)
// Nothing here invents a threshold.

import university2014 from "./data/university_2014.json";
import type {
  BandZh,
  FitnessItemId,
  Grade,
  GradeGroup,
  Sex,
} from "../types";

type ThresholdPair = [score: number, raw: number];

interface ItemDef {
  id: FitnessItemId;
  label_zh: string;
  label_en: string;
  unit: string;
  direction: "higher_is_better" | "lower_is_better" | "band_optimal";
  weight: number;
  scoring: "threshold" | "categorical";
  sex_specific_test?: boolean;
  thresholds?: Record<Sex, Record<GradeGroup, ThresholdPair[]>>;
  categories?: Array<{
    category: string;
    label_zh: string;
    score: number;
    male: { min: number | null; max: number | null };
    female: { min: number | null; max: number | null };
  }>;
}

interface BonusDef {
  basis: string;
  direction: "higher_is_better" | "lower_is_better";
  max_bonus: number;
  thresholds: Record<Sex, Record<GradeGroup, ThresholdPair[]>>;
}

interface TablesJson {
  standard: string;
  grade_band: string;
  score_levels: number[];
  grade_to_group: Record<Grade, GradeGroup>;
  item_bands: Record<BandZh, number[]>;
  items: ItemDef[];
  bonus: Record<string, BonusDef>;
  total_band_cutoffs: Record<BandZh, { total_min: number }>;
}

const TABLES = university2014 as unknown as TablesJson;

export function gradeToGroup(grade: Grade): GradeGroup {
  return TABLES.grade_to_group[grade];
}

export function itemMeta(item: FitnessItemId): ItemDef {
  const def = TABLES.items.find((i) => i.id === item);
  if (!def) throw new Error(`Unknown scoring item: ${item}`);
  return def;
}

/** Band (优秀/良好/...) for a single per-item score in [0,100]. */
export function bandForItemScore(score: number): BandZh {
  if (score >= 90) return "优秀";
  if (score >= 80) return "良好";
  if (score >= 60) return "及格";
  return "不及格";
}

/**
 * Score one raw measurement for an item.
 * Step function over the standard's printed [score, raw] pairs.
 * - higher_is_better: best score whose threshold raw <= value (vital capacity, jump, reach)
 * - lower_is_better:  best score whose threshold raw >= value (sprint, run time)
 * Values beyond the 100-point threshold earn a 0 per-item score here; 附加分 is
 * handled separately by bonusFor() so the bonus logic stays auditable.
 */
export function scoreItem(
  item: FitnessItemId,
  sex: Sex,
  grade: Grade,
  raw: number,
): number {
  const def = itemMeta(item);

  if (def.scoring === "categorical") {
    // Categories are anchored at their printed LOWER bounds, not matched on the
    // printed [min, max] interval. The standard prints bounds to one decimal
    // (低体重 <=17.8, 正常 17.9-23.9) but measured BMI is continuous, so a real
    // student at 17.8439 falls in none of the printed intervals. Anchoring at
    // lower bounds tiles the number line with no gap while preserving every
    // printed boundary. Port of aptams/rule_engine/tables.py::_score_categorical.
    const cats = def.categories ?? [];
    if (cats.length === 0) {
      throw new Error(`No categories defined for ${item}`);
    }
    const ordered = [...cats].sort(
      (a, b) =>
        (a[sex].min ?? Number.NEGATIVE_INFINITY) -
        (b[sex].min ?? Number.NEGATIVE_INFINITY),
    );
    let chosen: (typeof ordered)[number] | null = null;
    for (const cat of ordered) {
      const lo = cat[sex].min;
      if (lo === null || raw >= lo) chosen = cat;
    }
    if (!chosen) {
      throw new Error(`${item} ${raw} is below every category lower bound for ${sex}`);
    }
    return chosen.score;
  }

  const group = gradeToGroup(grade);
  const pairs = def.thresholds?.[sex]?.[group];
  if (!pairs || pairs.length === 0) {
    throw new Error(`No thresholds for ${item}/${sex}/${group}`);
  }

  if (def.direction === "higher_is_better") {
    // pairs are score-descending; raw thresholds also descending.
    let best = 0;
    for (const [s, r] of pairs) {
      if (raw >= r) {
        best = s;
        break;
      }
    }
    // Below the lowest (10-point) threshold -> 0.
    return best;
  }

  if (def.direction === "lower_is_better") {
    let best = 0;
    for (const [s, r] of pairs) {
      if (raw <= r) {
        best = s;
        break;
      }
    }
    return best;
  }

  throw new Error(`Unsupported direction for ${item}: ${def.direction}`);
}

/**
 * 附加分 (bonus). Awarded only by strength (reps above the 100-point threshold)
 * and endurance (seconds below the 100-point threshold). Capped per item at 10.
 * Returns the bonus points attributable to this item.
 */
export function bonusFor(
  item: FitnessItemId,
  sex: Sex,
  grade: Grade,
  raw: number,
): number {
  const bonusDef = TABLES.bonus[item];
  if (!bonusDef) return 0;

  const def = itemMeta(item);
  const group = gradeToGroup(grade);

  // The 100-point threshold is the first pair (score == 100) in the item table.
  const hundredPair = def.thresholds?.[sex]?.[group]?.find(([s]) => s === 100);
  if (!hundredPair) return 0;
  const hundredRaw = hundredPair[1];

  let excess: number;
  if (bonusDef.direction === "higher_is_better") {
    excess = raw - hundredRaw; // reps above 100-point mark
    if (excess <= 0) return 0;
  } else {
    excess = hundredRaw - raw; // seconds below 100-point mark
    if (excess <= 0) return 0;
  }

  // bonus thresholds are [bonus_points, excess] ascending in both.
  const bt = bonusDef.thresholds[sex][group];
  let points = 0;
  for (const [p, e] of bt) {
    if (excess >= e) points = p;
  }
  return Math.min(points, bonusDef.max_bonus);
}

export function itemWeight(item: FitnessItemId): number {
  return itemMeta(item).weight;
}

/**
 * The printed [score, raw] threshold pairs for an item/sex/grade group, in the
 * order stored in the standard (descending score). Each pair is an achievable
 * per-item score level and the raw value required to reach it.
 */
export function itemThresholds(
  item: FitnessItemId,
  sex: Sex,
  grade: Grade,
): Array<[number, number]> {
  const def = itemMeta(item);
  if (def.scoring === "categorical") return [];
  const group = gradeToGroup(grade);
  return def.thresholds?.[sex]?.[group] ?? [];
}

/** Bonus definition accessor (max bonus + direction) for a given item. */
export function itemBonusDef(item: FitnessItemId):
  | { max_bonus: number; direction: "higher_is_better" | "lower_is_better" }
  | undefined {
  const b = TABLES.bonus[item];
  return b ? { max_bonus: b.max_bonus, direction: b.direction } : undefined;
}

/**
 * TRUE while the total-level band cutoffs are a derived reading rather than a
 * printed table. The source standard groups per-ITEM scores under 等级 headings
 * (优秀 100-90, 良好 85-80, 及格 78-60, 不及格 50-10); applying that same
 * grouping to the weighted TOTAL is the conventional reading but is not printed
 * anywhere in the PDF. `data/scoring_tables/university_2014.json` marks it
 * `"derived": true` with "Confirm with mentors before any total-level band is
 * shown to a user" — so any UI showing a total band must mark it provisional.
 * Flip to false once mentors confirm (docs/task_a_results.md §4 item 5).
 */
export const TOTAL_BAND_IS_DERIVED = true;

/** Total weighted score -> grade band using the (derived) total cutoffs. */
export function bandForTotal(total: number): BandZh {
  const c = TABLES.total_band_cutoffs;
  if (total >= c["优秀"].total_min) return "优秀";
  if (total >= c["良好"].total_min) return "良好";
  if (total >= c["及格"].total_min) return "及格";
  return "不及格";
}

/**
 * The scholarship pass gate: evaluation-year total >= 60.0.
 *
 * RESOLVED FROM DATA, not assumed. Lowest passing score 60.0, highest failing
 * score 59.9, 100% agreement across all 20,140 students
 * (`analysis/route_to_pass.verify_pass_threshold`; docs/task_a_results.md §2).
 */
export const PASS_THRESHOLD = 60;

export const BAND_LABEL_EN: Record<BandZh, string> = {
  优秀: "Excellent",
  良好: "Good",
  及格: "Pass",
  不及格: "Fail",
};

export function allItemDefs(): ItemDef[] {
  return TABLES.items;
}

export function standardName(): string {
  return TABLES.standard;
}

File: reference-taskA/aptams/rule_engine/tables.py
"""Loader for the national-standard scoring tables.

The tables are a **blocking input** (``docs/proposal.md`` §11). They are not in the
repository and are never invented here. Until they are present in the configured directory
(``APTAMS_SCORING_TABLES_DIR``, default ``data/scoring_tables/``),
:func:`load_scoring_tables` raises :class:`ScoringTablesUnavailable`.

The on-disk JSON is produced by ``analysis/extract_scoring_tables.py``, which parses the
official PDF and aborts if the document's structure differs from what it expects. That
script plus the source PDF are the single auditable origin of every scored number in APTAMS
(``AGENTS.md`` §The core premise). Regenerate with::

    python -m analysis.extract_scoring_tables

This module is the only place a raw measurement becomes a score.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


class ScoringTablesUnavailable(RuntimeError):
    """Raised when scoring is attempted without the national-standard tables present.

    Deliberately a hard failure, not a fallback to approximate values. The ground-truth
    claim collapses the moment a threshold is guessed.
    """


class ScoringTableError(RuntimeError):
    """The tables are present but do not cover the requested lookup.

    Also a hard failure: an uncovered lookup means the standard says nothing about this
    case, and inventing an answer is exactly what the project forbids.
    """


def _default_dir() -> Path:
    return Path(os.environ.get("APTAMS_SCORING_TABLES_DIR", "data/scoring_tables"))


#: Score assigned when a raw value falls below the standard's lowest printed threshold.
#:
#: The source prints thresholds down to 10 points and stops. It does not state what a value
#: worse than the 10-point threshold scores. Zero is the only reading consistent with the
#: table being exhaustive downward, and it is confirmed empirically: it reproduces the
#: cohort's own ``total_score`` exactly (see ``analysis/verify_scorecard.py``). Recorded as a
#: named constant rather than a bare literal so the assumption stays visible and auditable.
BELOW_LOWEST_THRESHOLD_SCORE = 0.0


@dataclass(frozen=True)
class ScoringTables:
    """Parsed, validated scoring tables for one operative standard.

    Holds the per-item, per-sex, per-grade-group lookup structures plus item weightings,
    附加分 rules and band cutoffs, and exposes the read-only accessors used by
    :mod:`aptams.rule_engine.engine`. Every accessor either returns a value that came from
    the source document, or raises.
    """

    source_dir: Path
    standard: str
    grade_band: str
    items: dict[str, dict[str, Any]]
    bonus: dict[str, dict[str, Any]]
    grade_to_group: dict[str, str]
    total_band_cutoffs: dict[str, Any]
    raw: dict[str, Any] = field(repr=False, default_factory=dict)

    # --- introspection ---------------------------------------------------------------

    def item_ids(self) -> tuple[str, ...]:
        """Every scored item in the standard, in the source's own weighting order."""
        return tuple(self.items)

    def item(self, item_id: str) -> dict[str, Any]:
        try:
            return self.items[item_id]
        except KeyError:
            raise ScoringTableError(
                f"Item {item_id!r} is not in the loaded standard. Known items: "
                f"{sorted(self.items)}"
            ) from None

    def item_weight(self, *, item: str, band: str | None = None) -> float:
        """The item's weight as a fraction of the total (e.g. 0.20 for 50m sprint).

        ``band`` is accepted for interface compatibility; this standard publishes one
        weighting for the whole 大学 band.
        """
        if band is not None and band != self.grade_band:
            raise ScoringTableError(
                f"Loaded standard covers band {self.grade_band!r}, not {band!r}."
            )
        return float(self.item(item)["weight"]) / 100.0

    def grade_group(self, grade: str) -> str:
        """Map a study year (``g1``..``g4``) to the standard's grade group (大一大二/大三大四)."""
        try:
            return self.grade_to_group[grade]
        except KeyError:
            raise ScoringTableError(
                f"Unknown grade {grade!r}; the standard covers {sorted(self.grade_to_group)}"
            ) from None

    # --- scoring ---------------------------------------------------------------------

    def _thresholds(self, *, item: str, sex: str, grade: str) -> list[list[float]]:
        spec = self.item(item)
        if spec["scoring"] != "threshold":
            raise ScoringTableError(f"Item {item!r} is not threshold-scored.")
        group = self.grade_group(grade)
        try:
            return spec["thresholds"][sex][group]
        except KeyError:
            raise ScoringTableError(
                f"No thresholds for {item!r} / sex={sex!r} / group={group!r}."
            ) from None

    def score_item(
        self, *, item: str, sex: str, grade: str, raw: float, band: str | None = None
    ) -> float:
        """Score one raw measurement, straight from the table. No interpolation.

        The standard is a step function: a value earns the highest printed score whose
        threshold it meets. Intermediate values do **not** interpolate — a male first-year
        with 14 pull-ups scores 76, exactly as printed, because 80 requires 15.
        """
        if band is not None and band != self.grade_band:
            raise ScoringTableError(
                f"Loaded standard covers band {self.grade_band!r}, not {band!r}."
            )
        spec = self.item(item)
        if spec["scoring"] == "categorical":
            return self._score_categorical(spec, sex=sex, raw=raw)

        pairs = self._thresholds(item=item, sex=sex, grade=grade)
        higher_better = spec["direction"] == "higher_is_better"
        best = BELOW_LOWEST_THRESHOLD_SCORE
        for score, threshold in pairs:
            meets = raw >= threshold if higher_better else raw <= threshold
            if meets:
                best = max(best, float(score))
        return best

    @staticmethod
    def _score_categorical(spec: dict[str, Any], *, sex: str, raw: float) -> float:
        """BMI-style scoring: the raw value falls in exactly one category.

        The source prints category bounds to one decimal (低体重 ≤17.8, 正常 17.9~23.9, …),
        but measured BMI is continuous, so the printed bounds leave gaps — a real student at
        BMI 17.8439 sits in none of them. Categories are therefore anchored at their printed
        **lower** bounds: a value belongs to the last category whose lower bound it reaches.
        That tiles the continuous line with no gap and no rounding ambiguity, and preserves
        every printed boundary exactly. Verified against the cohort's own recorded totals.
        """
        ordered = sorted(
            spec["categories"],
            key=lambda c: (
                c[sex]["min"] if c[sex]["min"] is not None else float("-inf")
            ),
        )
        chosen = None
        for cat in ordered:
            lo = cat[sex]["min"]
            if lo is None or raw >= lo:
                chosen = cat
        if chosen is None:
            raise ScoringTableError(
                f"Value {raw} for {spec['id']!r} (sex={sex}) precedes every printed category."
            )
        return float(chosen["score"])

    def threshold_for_score(
        self, *, item: str, sex: str, grade: str, score: float
    ) -> float | None:
        """The raw value the standard requires for exactly ``score``.

        Returns ``None`` where the standard prints no threshold at that score (male pull-ups
        skip 78/74/70/66/62 — those scores are simply unreachable for that item). Used by
        Route-to-Pass to price the next reachable step.
        """
        for s, threshold in self._thresholds(item=item, sex=sex, grade=grade):
            if float(s) == float(score):
                return float(threshold)
        return None

    def reachable_scores(self, *, item: str, sex: str, grade: str) -> tuple[float, ...]:
        """The score levels this item can actually take, ascending. Excludes unprinted ones."""
        pairs = self._thresholds(item=item, sex=sex, grade=grade)
        return tuple(sorted(float(s) for s, _ in pairs))

    # --- 附加分 ------------------------------------------------------------------------

    def bonus_items(self) -> tuple[str, ...]:
        return tuple(self.bonus)

    def bonus_for(self, *, item: str, sex: str, grade: str, raw: float) -> float:
        """附加分 earned on ``item``, per the source's 加分指标评分表.

        Strength is 高优: reps *above* the 100-point threshold earn bonus. Endurance is
        低优: seconds *below* it earn bonus. Items with no bonus table earn nothing.
        """
        spec = self.bonus.get(item)
        if spec is None:
            return 0.0
        hundred = self.threshold_for_score(item=item, sex=sex, grade=grade, score=100)
        if hundred is None:
            raise ScoringTableError(f"No 100-point threshold for {item!r}; cannot price bonus.")
        margin = (raw - hundred) if spec["direction"] == "higher_is_better" else (hundred - raw)
        if margin <= 0:
            return 0.0
        group = self.grade_group(grade)
        try:
            pairs = spec["thresholds"][sex][group]
        except KeyError:
            raise ScoringTableError(
                f"No bonus table for {item!r} / sex={sex!r} / group={group!r}."
            ) from None
        earned = 0.0
        for points, needed in pairs:
            if margin >= needed:
                earned = max(earned, float(points))
        return min(earned, float(spec["max_bonus"]))

    # --- bands -------------------------------------------------------------------------

    def grade_band_for_total(self, *, total: float, band: str | None = None) -> str:
        """The 等级 label for a weighted total.

        Note these cutoffs are marked ``derived`` in the extracted tables: the source prints
        等级 groupings for per-*item* scores, not for the total. See the ``$note`` field.
        """
        if band is not None and band != self.grade_band:
            raise ScoringTableError(
                f"Loaded standard covers band {self.grade_band!r}, not {band!r}."
            )
        labelled = [
            (label, spec["total_min"])
            for label, spec in self.total_band_cutoffs.items()
            if isinstance(spec, dict) and "total_min" in spec
        ]
        for label, minimum in sorted(labelled, key=lambda kv: -kv[1]):
            if total >= minimum:
                return label
        raise ScoringTableError(f"No band covers total {total}.")

    @property
    def total_band_cutoffs_are_derived(self) -> bool:
        """True while the total-level cutoffs remain an inference awaiting mentor confirmation."""
        return bool(self.total_band_cutoffs.get("derived", False))


def _validate(payload: dict[str, Any], source_dir: Path) -> ScoringTables:
    """Validate the parsed JSON before any of it is trusted for scoring."""
    for key in ("standard", "grade_band", "items", "bonus", "grade_to_group"):
        if key not in payload:
            raise ScoringTableError(f"Scoring tables in {source_dir} lack required key {key!r}.")

    items = {spec["id"]: spec for spec in payload["items"]}
    if not items:
        raise ScoringTableError(f"Scoring tables in {source_dir} contain no items.")

    total_weight = sum(float(spec["weight"]) for spec in items.values())
    if round(total_weight, 6) != 100.0:
        raise ScoringTableError(
            f"Item weights sum to {total_weight}, not 100 — refusing to score with a "
            "weighting that does not match the standard."
        )
    for item_id, spec in items.items():
        if spec["scoring"] not in {"threshold", "categorical"}:
            raise ScoringTableError(f"Item {item_id!r} has unknown scoring {spec['scoring']!r}.")

    return ScoringTables(
        source_dir=source_dir,
        standard=payload["standard"],
        grade_band=payload["grade_band"],
        items=items,
        bonus=payload.get("bonus", {}),
        grade_to_group=payload["grade_to_group"],
        total_band_cutoffs=payload.get("total_band_cutoffs", {}),
        raw=payload,
    )


def load_scoring_tables(source_dir: str | os.PathLike[str] | None = None) -> ScoringTables:
    """Load and validate the scoring tables from ``source_dir``.

    Raises :class:`ScoringTablesUnavailable` if the directory holds no table data — the T0
    state, and the state any teammate is in before running the extractor.
    """
    path = Path(source_dir) if source_dir is not None else _default_dir()

    table_files = [
        p
        for p in sorted(path.glob("*"))
        if p.is_file()
        and p.suffix.lower() in {".json", ".csv", ".xlsx", ".yaml", ".yml"}
        and not p.name.endswith(".example.json")
    ]
    if not table_files:
        raise ScoringTablesUnavailable(
            f"No scoring tables found in {path!s}. The 《大学生国家体质健康测试评分标准》 tables "
            "are a blocking input and are never guessed. Generate them from the source PDF "
            "with:  python -m analysis.extract_scoring_tables"
        )
    if len(table_files) > 1:
        raise ScoringTableError(
            f"Expected exactly one scoring-table file in {path!s}, found "
            f"{[p.name for p in table_files]}. Ambiguous tables are refused rather than "
            "silently picking one."
        )

    payload = json.loads(table_files[0].read_text(encoding="utf-8"))
    return _validate(payload, path)
